### PR TITLE
Add internal switch

### PR DESCRIPTION
**Pull request by AR-May**
Changes in this PR:
- Adding --internal switch for aka.ms links.
- Fixing usage of the credentials for aka.ms links (previously credentials were applied twice and as a result the final link had wrong format).
- Prevent echoing of credentials given --feed-credential switch.

Delivers issue #169.

### CODE REVIEW

File: src/dotnet-install.ps1 [hunk 1]
```diff
@@ -837,17 +846,19 @@ function Get-AkaMSDownloadLink([string]$Channel, [string]$Quality, [string]$Prod
...
     if (-not [string]::IsNullOrEmpty($Quality)) {
         $akaMsLink +="/$Quality"
     }
-
     $akaMsLink +="/$Product-win-$Architecture.zip"
     Say-Verbose  "Constructed aka.ms link: '$akaMsLink'."
 
     #get HTTP response
     $Response= GetHTTPResponse -Uri $akaMsLink -HeaderOnly $true -DisableRedirect $true
-    Say-Verbose "Received response:`n$Response"
```
Comment (AR-May) [inline src/dotnet-install.ps1:850]: Response has credentials in the redirect link, we do not want to log it.

File: src/dotnet-install.ps1 [hunk 2]
```diff
@@ -859,6 +870,9 @@ function Get-AkaMSDownloadLink([string]$Channel, [string]$Quality, [string]$Prod
     {
         try {
             $akaMsDownloadLink = $Response.Headers.GetValues("Location")[0]
+            #FeedCredential is the part of redirect link, remove it
+            $akaMsDownloadLink = $akaMsDownloadLink.split('?')[0]
```
Comment (AR-May) [inline src/dotnet-install.ps1:874]: I removed the credentials from the redirect link here to avoid double application of credentials while the actual downloading. However, might be better to not apply credentials to the aka.ms link in the first place, i.e. have a bool parameter with the meaning if it is needed to apply credentials or not in GetHTTPResponse function.

File: src/dotnet-install.ps1 [hunk 3]
```diff
@@ -32,6 +32,8 @@
     - latest - most latest build on specific channel
     - 3-part version in a format A.B.C - represents specific version of build
           examples: 2.0.0-preview2-006120, 1.1.0
+.PARAMETER Internal
+    Set this parameter to access internal builds. It might be nessesary also to use FeedCredential parameter with it.
```
Comment (bozturkMSFT) [inline src/dotnet-install.ps1:36]: My suggestion:
Set this switch to download internal builds. It is necessary to use FeedCredential option with this.

File: src/dotnet-install.ps1 (hunk 1 above)
Comment (bozturkMSFT) [reply src/dotnet-install.ps1:850]: Would it be useful to know that the response was received? If yes, maybe we can just log "Response received." without printing the content.

Comment (AR-May) [reply src/dotnet-install.ps1:850]: Yes, but this line actually will be back when I remove credentials from aka.ms link, it would be safe then.

File: src/dotnet-install.ps1 [hunk 4]
```diff
@@ -122,6 +125,12 @@ if ($SharedRuntime -and (-not $Runtime)) {
     $Runtime = "dotnet"
 }
 
+#FeedCredential should start with "?", for it to be added to the end of the link.
+#adding "?" at the beginning of the FeedCredential if needed.
+if ((![string]::IsNullOrEmpty($FeedCredential)) -and ($FeedCredential[0] -ne '?')) {
```
Comment (donJoseLuis) [inline src/dotnet-install.ps1:130]: minor: are there other restrictions for the `FeedCredential` string?  If yes, does it make sense to also enforce ithere?

File: tests/Install-Scripts.Test/AkaMsLinksTests.cs [hunk 5]
```diff
@@ -277,6 +277,85 @@ public void LinkCanBeCreatedForGivenRuntime(string channel, string runtime, stri
             commandResult.Should().HaveStdOutContaining(output => Regex.IsMatch(output, expectedLinkPattern));
         }
 
+        [Theory]
+        [InlineData("3.1", null, @"https://aka.ms/dotnet/internal/3.1/dotnet-sdk-")]
+        [InlineData("5.0.2xx", null, @"https://aka.ms/dotnet/internal/5.0.2xx/dotnet-sdk-")]
+        [InlineData("Current", null, @"https://aka.ms/dotnet/internal/current/dotnet-sdk-")]
+        [InlineData("LTS", null, @"https://aka.ms/dotnet/internal/LTS/dotnet-sdk-")]
+        [InlineData("5.0.2xx", "validated", @"https://aka.ms/dotnet/internal/5.0.2xx/validated/dotnet-sdk-")]
+        public void InternalLinkCanBeCreatedForSdk(string channel, string quality, string expectedLink)
```
Comment (donJoseLuis) [inline tests/Install-Scripts.Test/AkaMsLinksTests.cs:286]: consider consolidating content from both tests, they share a lot of the same exact code.

File: tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs [hunk 6]
```diff
@@ -8,11 +8,14 @@
 using System.Collections.Generic;
 using Microsoft.NET.TestFramework.Assertions;
 using FluentAssertions;
+using System.Text.RegularExpressions;
 
 namespace Microsoft.DotNet.InstallationScript.Tests
 {
     public class GivenThatIWantToGetTheSdkLinksFromAScript : TestBase
     {
+        private const string _feedCredential = "478a920c-2217-49f2-9c31-2fc3b4fef7cb";
```
Comment (donJoseLuis) [inline tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs:17]: please add comments describing the credential.  valid credentials should not be enlisted.  No problem if this is a string, without any access or authorization, merely used for testing.

**Review by donJoseLuis — CHANGES_REQUESTED** [review]

File: src/dotnet-install.ps1 [hunk 7]
```diff
@@ -297,7 +306,7 @@ function Load-Assembly([string] $Assembly) {
     }
 }
 
-function GetHTTPResponse([Uri] $Uri, [bool]$HeaderOnly, [bool]$DisableRedirect)
+function GetHTTPResponse([Uri] $Uri, [bool]$HeaderOnly, [bool]$DisableRedirect, [bool]$DisableFeedCredential)
```
Comment (bozturkMSFT) [inline src/dotnet-install.ps1:309]: Does the new parameter default to `false` when not specified?

File: src/dotnet-install.sh [hunk 8]
```diff
@@ -869,7 +869,7 @@ extract_dotnet_package() {
 }
 
 # args:
-# remote_path - $1
+# remote_path - $1 - Note, credentials will not be applied.
```
Comment (bozturkMSFT) [inline src/dotnet-install.sh:872]: As I understand, we are still passing the credentials to aka.ms links in bash script like: `aka.ms/dotnet/internal/daily/some_archive.zip?credentials`. 
But in powershell script, aka.ms doesn't get the credentials. Is it intended that they are different?

File: tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs [hunk 9]
```diff
@@ -91,6 +94,32 @@ public void WhenRuntimeParametersArePassedToInstallScripts(string runtime, strin
             commandResult.Should().HaveStdOutContainingIgnoreCase(runtimeType);
         }
 
+
+        [Theory]
+        [InlineData("3.1", _feedCredential)]
+        [InlineData("5.0.2xx", _feedCredential)]
+        [InlineData("LTS", "?" + _feedCredential)]
+        [InlineData("release/1.0.0", "?" + _feedCredential)]
```
Comment (bozturkMSFT) [inline tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs:102]: These have extra `?`. Doesn't this break the test?

File: tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs [hunk 10]
```diff
@@ -91,6 +94,32 @@ public void WhenRuntimeParametersArePassedToInstallScripts(string runtime, strin
...
+                            .Execute();
+
+            //  Standard 'dryrun' criterium
+            commandResult.Should().Pass();
+            commandResult.Should().NotHaveStdOutContaining("dryrun");
+            commandResult.Should().HaveStdOutContaining("Repeatable invocation:");
+
+            // Checking format of the link after applying feed credential
+            string linkWithCredentialsExpression = @"Primary named payload URL: [a-zA-Z0-9:/.-]*\?" + feedCredential;
+            commandResult.Should().HaveStdOutContaining(output => Regex.IsMatch(output, linkWithCredentialsExpression));
```
Comment (bozturkMSFT) [inline tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs:120]: We have two links printed, is that right? One for primary and one for legacy. If legacy one works and primary one is broken, the test will pass today. It would be nice if both were checked separately.

File: src/dotnet-install.ps1 (hunk 7 above)
Comment (AR-May) [reply src/dotnet-install.ps1:309]: yes, the same as for the other parameters. I have checked that, sure.

File: tests/Install-Scripts.Test/AkaMsLinksTests.cs (hunk 5 above)
Comment (AR-May) [reply tests/Install-Scripts.Test/AkaMsLinksTests.cs:286]: I discussed this with @bozturkMSFT and the current structure is that we keep use cases for sdk and for runtime separate in unit tests. @bozturkMSFT, could you please comment on the reasons for that?

File: src/dotnet-install.ps1 (hunk 4 above)
Comment (AR-May) [reply src/dotnet-install.ps1:130]: The structure of FeedCredential string is unfortunately something that I have little information about. Thinking about that, behind the aka.ms link could be different storages, and it might be that credential strings for them could vary in format. They definitely has to be passed as parameters within, so that's why we need to add them with "?" to the URI.

File: tests/Install-Scripts.Test/AkaMsLinksTests.cs (hunk 5 above)
Comment (AR-May) [reply tests/Install-Scripts.Test/AkaMsLinksTests.cs:286]: I'd say that we do not want to have too many ifs in the unit test (checking whether we have runtime specified in the command), and do not want to confuse ourselves with different parameters that we pass to the test and their possible combinations. Maybe there are also other reasons for that.

File: tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs (hunk 9 above)
Comment (AR-May) [reply tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs:102]: No, it does not. The previous behavior is that credentials should have "?" in the beginning. I added an opportunity to pass credentials without it, but did not change the previous behavior, so credentials still could be starting "?", for the backward compatibility. Now you can pass it either way, with or without "?".

File: src/dotnet-install.sh (hunk 8 above)
Comment (AR-May) [reply src/dotnet-install.sh:872]: The code structure related to downloading differs for sh and ps1 scripts. So, removal of the credentials from aka.ms links was done differently.
So, no, we are not adding credentials to aka.ms link, I removed that from this function, you can see it below. The comment actually indicates this fact. We no more apply credentials in this function, which is used only for resolution of aka.ms links into a direct ones.

Comment (AR-May) [reply src/dotnet-install.sh:872]: $remote_path_with_credential -> $remote_path is the related change, removing the credentials from aka.ms links. It is done in `get_http_header_curl` and `get_http_header_wget`. This function is just a generic one, it allows us not to care about what we have on the machine for downloading, wget or curl.

File: tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs (hunk 10 above)
Comment (AR-May) [reply tests/Install-Scripts.Test/GivenThatIWantToGetTheSdkLinksFromAScript.cs:120]: We do not always have two links printed, the guaranteed one is the primary one. The legacy one is printed if it is not empty and it happens sometimes. So, we need either to copy the behavior of the scripts and compute a condition indicating whether we have the legacy link, or do not put much of an effort here and just check only the primary link. I think the check of the primary link with the credentials is currently enough.

File: tests/Install-Scripts.Test/AkaMsLinksTests.cs (hunk 5 above)
Comment (bozturkMSFT) [reply tests/Install-Scripts.Test/AkaMsLinksTests.cs:286]: The reason is not to complicate the logic in the test methods.

File: src/dotnet-install.ps1 [hunk 11]
```diff
@@ -901,6 +895,16 @@ Say "- The SDK needs to be installed without user interaction and without admin
 Say "- The SDK installation doesn't need to persist across multiple CI runs."
 Say "To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.`r`n"
 
+if ($Internal -and [string]::IsNullOrEmpty($FeedCredential)) {
+    Say-Warning "Note, it is necessary to use FeedCredential option when Internal switch is set."
```
Comment (bozturkMSFT) [inline src/dotnet-install.ps1:899]: I think this should start with "Note that..."

File: src/dotnet-install.sh [hunk 12]
```diff
@@ -1481,6 +1481,10 @@ say "- The SDK needs to be installed without user interaction and without admin
 say "- The SDK installation doesn't need to persist across multiple CI runs."
 say "To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.\n"
 
+if [ "$internal" = true ] && [ -z "$feed_credential" ]; then
+    say_warning "Note, it is necessary to use --feed-credential option when --internal switch is set."
```
Comment (bozturkMSFT) [inline src/dotnet-install.sh:1485]: We have `--` before the switches but not in powershell. Should we make them the same in both scripts?

**Review by bozturkMSFT — APPROVED** [review]

File: src/dotnet-install.ps1 [hunk 13]
```diff
@@ -32,6 +32,8 @@
     - latest - most latest build on specific channel
     - 3-part version in a format A.B.C - represents specific version of build
           examples: 2.0.0-preview2-006120, 1.1.0
+.PARAMETER Internal
+    Set this switch to download internal builds. It is necessary to use FeedCredential option with this.
```
Comment (vlada-shubina) [inline src/dotnet-install.ps1:36]: nit: imo this wording is not common. Should we use "Applicable only when FeedCredential option is set"?

File: src/dotnet-install.ps1 (hunk 11 above)
Comment (vlada-shubina) [reply src/dotnet-install.ps1:899]: I would reword as "Applicable only when FeedCredential option is set"

File: src/dotnet-install.ps1 (hunk 13 above)
Comment (vlada-shubina) [reply src/dotnet-install.ps1:36]: Should it be put near FeedCredential?

File: src/dotnet-install.ps1 [hunk 14]
```diff
@@ -938,9 +963,9 @@ $ScriptName = $MyInvocation.MyCommand.Name
 
 if ($DryRun) {
     Say "Payload URLs:"
-    Say "Primary named payload URL: $DownloadLink"
+    Say "Primary named payload URL: ${DownloadLink}${FeedCredential}"
```
Comment (vlada-shubina) [inline src/dotnet-install.ps1:966]: I don't think we are allowed to print feed credentials.

File: tests/Install-Scripts.Test/AkaMsLinksTests.cs (hunk 5 above)
Comment (vlada-shubina) [inline tests/Install-Scripts.Test/AkaMsLinksTests.cs:286]: I think 'internal' can be just a parameter to existing test methods.

**Review by vlada-shubina — CHANGES_REQUESTED** [review]
Please review adding the feed credentials to log due a security reason.

File: src/dotnet-install.ps1 (hunk 14 above)
Comment (AR-May) [reply src/dotnet-install.ps1:966]: This is not showed in logs, even verbose, this is the dry-run. FeedCredential is already logged in dry run few lines below in the repeatable command. It makes sense to me to have FeedCredential as a part of URL here, so the person can actually have a downloadable link. But this may be debatable.

File: src/dotnet-install.ps1 (hunk 13 above)
Comment (AR-May) [reply src/dotnet-install.ps1:36]: This parameter is quite similar to "channel", "version" and "quality" (as it also describes what we would like to install), so I put it near them. I can move "FeedCredential" near them also, though.

File: tests/Install-Scripts.Test/AkaMsLinksTests.cs (hunk 5 above)
Comment (AR-May) [reply tests/Install-Scripts.Test/AkaMsLinksTests.cs:286]: Well, I can unite these tests.

File: src/dotnet-install.ps1 (hunk 11 above)
Comment (AR-May) [reply src/dotnet-install.ps1:899]: So, summarizing: "Note that Internal switch is applicable only when FeedCredential option is set."? To me it sounds a bit like the switch would be ignored otherwise, which is not the case, we will get a download error.  But this is still better, yes.

File: src/dotnet-install.sh (hunk 12 above)
Comment (nohwnd) [reply src/dotnet-install.sh:1485]: Probably not because 1) powershell code won't work with --Internal, and 2) -internal in sh would mean that `-i -n -t -e -r -n -a -l` parameters are used in sh?

File: src/dotnet-install.ps1 (hunk 13 above)
Comment (nohwnd) [reply src/dotnet-install.ps1:36]: I don't agree fully with the new wording either. 
```suggestion
    Download internal builds. Provide credentials via -FeedCredential parameter.
```

o

Comment (AR-May) [reply src/dotnet-install.ps1:36]: Yes, I like your version more, @nohwnd, thank you!

File: src/dotnet-install.ps1 (hunk 11 above)
Comment (nohwnd) [reply src/dotnet-install.ps1:899]: I would just state that assumption:

```suggestion
    Say-Warning "Command was called with -Internal parameter, but no credentials were provided using -FeedCredential parameter. You will most likely get authorization errors."
```

File: src/dotnet-install.sh (hunk 12 above)
Comment (AR-May) [reply src/dotnet-install.sh:1485]: I suppose @bozturkMSFT means that I missed "-" before the switch name in the similar message in the ps script. I will add it.

File: src/dotnet-install.ps1 [hunk 15]
```diff
@@ -880,6 +895,20 @@ Say "- The SDK needs to be installed without user interaction and without admin
 Say "- The SDK installation doesn't need to persist across multiple CI runs."
 Say "To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.`r`n"
 
+if ($Internal -and [string]::IsNullOrEmpty($FeedCredential)) {
```
Comment (donJoseLuis) [inline src/dotnet-install.ps1:898]: minor:  should validation pass if the feed credential is white space?  Such as `"     "`.

File: src/dotnet-install.ps1 [hunk 16]
```diff
@@ -880,6 +895,20 @@ Say "- The SDK needs to be installed without user interaction and without admin
...
+    if ($DryRun) {
+        Say-Warning "Provide credentials via -FeedCredential parameter."
```
Comment (donJoseLuis) [inline src/dotnet-install.ps1:900]: minor: it's possible these strings can get out of sync.  Should the string be declared once & then be used in both `Say-Warning` & `throw`?

File: src/dotnet-install.ps1 (hunk 14 above)
Comment (donJoseLuis) [reply src/dotnet-install.ps1:970]: Proper general guidance by @vlada-shubina .  In this dry-run's use case, the URL should be actionable.

**Review by donJoseLuis — APPROVED** [review]

File: src/dotnet-install.ps1 (hunk 15 above)
Comment (AR-May) [reply src/dotnet-install.ps1:898]: Yes, it is better if we consider this case and strip the input. I will fix this.

File: src/dotnet-install.ps1 [hunk 17]
```diff
@@ -880,6 +895,22 @@ Say "- The SDK needs to be installed without user interaction and without admin
 Say "- The SDK installation doesn't need to persist across multiple CI runs."
 Say "To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.`r`n"
 
+$FeedCredential = $FeedCredential.Trim();
```
[the remark this answers is not in the document]
Comment (AR-May) [reply src/dotnet-install.ps1:898]: Bash version behaves the same. Ok, I can check if it `IsNullOrWhitespace` instead.

File: src/dotnet-install.sh [hunk 18]
```diff
@@ -1408,6 +1439,8 @@ do
             echo "          For SDK use channel in A.B.Cxx format. Using quality for SDK together with channel in A.B format is not supported." 
             echo "          Supported since 5.0 release." 
             echo "          Note: The version parameter overrides the channel parameter when any version other than `latest` is used, and therefore overrides the quality."
+            echo "  --internal,-Internal               Download internal builds. Provide credentials via --feed-credential parameter."
+            echo "  --feed-credential,-FeedCredential  Azure feed shared access token. This parameter typically is not specified."
```
[the remark this answers is not in the document]
Comment (AR-May) [reply src/dotnet-install.sh:1443]: The more general version with `Token to access Azure feed.` sounds better, yes.

File: src/dotnet-install.sh [hunk 19]
```diff
@@ -1469,28 +1501,45 @@ say "- The SDK needs to be installed without user interaction and without admin
 say "- The SDK installation doesn't need to persist across multiple CI runs."
 say "To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.\n"
 
+if [ "$internal" = true ] && [ -z "$feed_credential" ]; then
+    message="Provide credentials via --feed-credential parameter."
+    if [ "$dry_run" = true ]; then
```
[the remark this answers is not in the document]
Comment (AR-May) [reply src/dotnet-install.sh:1506]: Well, it is less code if we always fail, right, but to me --dry-run should not fail in this case, as it really does not need the access to the resource.

File: src/dotnet-install.sh [hunk 20]
```diff
@@ -870,16 +870,19 @@ extract_dotnet_package() {
...
 {
     eval $invocation
     local remote_path="$1"
+    local disable_feed_credential="$2"
+
     local failed=false
     local response
     if machine_has "curl"; then
-        get_http_header_curl $remote_path || failed=true
+        get_http_header_curl $remote_path "$disable_feed_credential" || failed=true
```
[the remark this answers is not in the document]
Comment (AR-May) [reply src/dotnet-install.sh:883]: Seems like they are not needed, right.

**Review by vlada-shubina — CHANGES_REQUESTED** [review]
some comments about tests worth attention

File: src/dotnet-install.ps1 [hunk 21]
```diff
@@ -33,7 +33,10 @@
     - 3-part version in a format A.B.C - represents specific version of build
           examples: 2.0.0-preview2-006120, 1.1.0
 .PARAMETER Internal
-    Set this switch to download internal builds. It is necessary to use FeedCredential option with this.
+    Download internal builds. Requires providing credentials via -FeedCredential parameter.
+.PARAMETER FeedCredential
+    Token to access Azure feed. Used as a query string to append to the Azure feed.
+    It allows changing the URL to use non-public blob storage accounts. This parameter typically is not specified.
```
[the remark this answers is not in the document]
Comment (bozturkMSFT) [reply src/dotnet-install.ps1:39]: This one is a reference to `AzureFeed` option. You change your feed to something else using `AzureFeed`, then you set `FeedCredential` to get access to that feed.

Though I agree that it can use a bit of clarification. Maybe something like:
```
It allows access to the private feed specified by "-AzureFeed" option in the case that the default feed is not used.
```

File: src/dotnet-install.ps1 [hunk 22]
```diff
@@ -859,6 +859,9 @@ function Get-AkaMSDownloadLink([string]$Channel, [string]$Quality, [bool]$Intern
     Say-Verbose  "Constructed aka.ms link: '$akaMsLink'."
 
     #get HTTP response
+    #do not pass credentials as a part of the $akaMsLink and do not apply credentials in the GetHTTPResponse function
```
[the remark this answers is not in the document]
Comment (bozturkMSFT) [reply src/dotnet-install.ps1:862]: Yes. Aka.ms links are public redirections. The credentials are needed in the resolved URL only.

File: src/dotnet-install.sh (hunk 19 above)
[the remark this answers is not in the document]
Comment (bozturkMSFT) [reply src/dotnet-install.sh:1510]: DryRun still downloads files to determine the latest version of that channel (when used with channels that don't support aka.ms links). They will see a 404 error in that case because the credentials are not there. I think it is better to prevent that scenario.

File: tests/Install-Scripts.Test/GivenThatIWantToInstallDotnetFromAScript.cs [hunk 23]
```diff
@@ -312,13 +315,92 @@ public void WhenInstallingWindowsdesktopRuntime(string channel, string? quality,
...
 
-        private static IEnumerable<string> GetInstallScriptArgs(string? channel, string? runtime, string? quality, string? installDir)
+        [Theory]
+        [MemberData(nameof(InstallSdkFromChannelTestCases))]
+        public void WhenInstallingTheSdkWithFeedCredential(string channel, string? quality, string versionRegex)
+        {
+            string feedCredential = "?" + Guid.NewGuid().ToString();
+
+            // Run install script to download and install.
+            var args = GetInstallScriptArgs(channel, null, quality, _sdkInstallationDirectory, feedCredential, true);
```
Comment (vlada-shubina) [inline tests/Install-Scripts.Test/GivenThatIWantToInstallDotnetFromAScript.cs:332]: still nameless `true` here and in test below

**Review by vlada-shubina — APPROVED** [review]

File: src/dotnet-install.sh [hunk 24]
```diff
@@ -1359,7 +1394,9 @@ do
         --feed-credential|-[Ff]eed[Cc]redential)
             shift
             feed_credential="$1"
-            non_dynamic_parameters+=" $name "\""$1"\"""
+            #feed_credential should start with "?", for it to be added to the end of the link.
+            #adding "?" at the beginning of the feed_credential if needed.
+            [[ -z "$(echo $feed_credential)" ]] || [[ $feed_credential == \?* ]] || feed_credential="?$feed_credential"
```
Comment (ladipro) [inline src/dotnet-install.sh:1399]: Curious why the `echo` is needed. Couldn't we just do:
```suggestion
            [[ -z "$feed_credential" ]] || [[ $feed_credential == \?* ]] || feed_credential="?$feed_credential"
```

File: src/dotnet-install.sh [hunk 25]
```diff
@@ -1469,28 +1508,45 @@ say "- The SDK needs to be installed without user interaction and without admin
 say "- The SDK installation doesn't need to persist across multiple CI runs."
 say "To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.\n"
 
+if [ "$internal" = true ] && [ -z "$(echo $feed_credential)" ]; then
```
Comment (ladipro) [inline src/dotnet-install.sh:1511]: Same question about `echo`. It looks redundant.

File: src/dotnet-install.sh [hunk 26]
```diff
@@ -1469,28 +1508,45 @@ say "- The SDK needs to be installed without user interaction and without admin
...
-        say "Legacy named payload URL: $legacy_download_link"
+        say "Legacy named payload URL: ${legacy_download_link}"
     fi
     repeatable_command="./$script_name --version "\""$specific_version"\"" --install-dir "\""$install_root"\"" --architecture "\""$normalized_architecture"\"" --os "\""$normalized_os"\"""
     
     if [ ! -z "$normalized_quality" ]; then
         repeatable_command+=" --quality "\""$normalized_quality"\"""
     fi
 
+
```
Comment (ladipro) [inline src/dotnet-install.sh:1537]: super-nit: Was this empty line added unintentionally?

**Review by ladipro — APPROVED** [review]
Still looking good. A couple of nits inline, definitely not blocking.

File: src/dotnet-install.sh (hunk 24 above)
Comment (AR-May) [reply src/dotnet-install.sh:1399]: This is to have the same behavior as .ps1 script. In powershell we have 
`[string]::IsNullOrWhitespace($FeedCredential)`.  
To do the same in bash we trim the value with echo and then check if it is an empty string.

File: src/dotnet-install.sh (hunk 26 above)
Comment (AR-May) [reply src/dotnet-install.sh:1537]: That's unintentional.

File: src/dotnet-install.sh (hunk 24 above)
Comment (ladipro) [reply src/dotnet-install.sh:1399]: Neat! I didn't know about this new trick, thank you.

File: src/dotnet-install.ps1 (hunk 21 above)
[the remark this answers is not in the document]
Comment (AR-May) [reply src/dotnet-install.ps1]: I am not sure I would like to do more rearrangements in the parameters order, there is also `UncachedFeed` that we would like to have near `AzureFeed`. Maybe this might lead to even more problems with order.
